src/Psalm/Plugin/ArgTypeInferer.php

```diff
@@ -0,0 +1,36 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psalm\Plugin;
+
+use PhpParser;
+use Psalm\Context;
+use Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer;
+use Psalm\Internal\Analyzer\StatementsAnalyzer;
+use Psalm\Type;
+use Psalm\Type\Union;
+
+final class ArgTypeInferer
+{
+    private Context $context;
+    private StatementsAnalyzer $statements_analyzer;
+
+    public function __construct(Context $context, StatementsAnalyzer $statements_analyzer)
+    {
+        $this->context = $context;
+        $this->statements_analyzer = $statements_analyzer;
+    }
+
+    /**
+     * @return false|Union
+     */
+    public function infer(PhpParser\Node\Arg $arg)
+    {
+        if (ExpressionAnalyzer::analyze($this->statements_analyzer, $arg->value, $this->context) === false) {
```

Comment:
It might be possible that the expression was already analysed - for example inside the `FunctionReturnTypeProvider` callbacks.

Might it be a better idea to first check if the statement analyzer already contains the argument before analysing it?

Reply:
Good catch!